File: backend-reto/src/main/java/com/backend/reto/service/CategoryService.java
package com.backend.reto.service;

import com.backend.reto.exception.ResourceNotFoundException;
import com.backend.reto.model.Category;
import com.backend.reto.repository.CategoryRepository;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class CategoryService {
    @Autowired
    private CategoryRepository categoryRepository;

    public List<Category> findAll() {
        return categoryRepository.findAll();
    }

    public Category findById(Long id) {
        return categoryRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Category not found with id: " + id));
    }

    public Category save(Category category) {
        return categoryRepository.save(category);
    }

    public void delete(Long id) {
        Category category = categoryRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Category not found"));

        // Verificar si la categoría tiene notas asociadas
        if (!category.getNotes().isEmpty()) {
            throw new IllegalStateException("No se puede eliminar una categoría que tiene notas asociadas");
        }

        categoryRepository.deleteById(id);
    }
}

File: frontend-reto/src/services/categoryService.js
import api from './api';

export const getAllCategories = () => api.get('/categories');
export const getCategoryById = (id) => api.get(`/categories/${id}`);
export const createCategory = (category) => api.post('/categories', category);
export const deleteCategory = async (id) => {
    try {
      const response = await api.delete(`/categories/${id}`);
      return response.data;
    } catch (error) {
      if (error.response?.status === 500) {
        throw new Error('No se puede eliminar una categoría que tiene notas asociadas');
      }
      throw error;
    }
  };
